### PR TITLE
Infinite scrolling in stats view all

### PR DESCRIPTION
**Pull request by planarvoid**
This PR adds infinite scrolling to the view all screens (with an example on Followers). How it's done: There is a `LoadingItem` that shows a progress bar in the list and when it's bound in the adapter, it triggers the loading of the next page. 

There is no explicit error handling yet. When the next page load fails, we show a full-screen error. That will be handled in another PR and might be a bit too complex. 

To test:
* Go to Stats/Insights
* Click on "View more"
* Scroll down
* More pages are automatically loaded

### CODE REVIEW

**Review by 0nko — CHANGES_REQUESTED** [review]
When I try to open Followers I'm seeing a crash:
```
2019-03-04 20:49:06.989 18752-18752/org.wordpress.android.beta E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.wordpress.android.beta, PID: 18752
    java.lang.IllegalStateException: Cannot call this method while RecyclerView is computing a layout or scrolling android.support.v7.widget.RecyclerView{e355768 VFED.V... ......ID 0,0-1080,1605 #7f090267 app:id/loadingRecyclerView}, adapter:org.wordpress.android.ui.stats.refresh.lists.sections.BlockListAdapter@fc81781, layout:android.support.v7.widget.LinearLayoutManager@e218026, context:org.wordpress.android.ui.stats.refresh.StatsViewAllActivity@5a8de90
        at android.support.v7.widget.RecyclerView.assertNotInLayoutOrScroll(RecyclerView.java:2880)
        at android.support.v7.widget.RecyclerView$RecyclerViewDataObserver.onItemRangeRemoved(RecyclerView.java:5308)
        at android.support.v7.widget.RecyclerView$AdapterDataObservable.notifyItemRangeRemoved(RecyclerView.java:12032)
        at android.support.v7.widget.RecyclerView$Adapter.notifyItemRangeRemoved(RecyclerView.java:7248)
        at android.support.v7.util.AdapterListUpdateCallback.onRemoved(AdapterListUpdateCallback.java:48)
        at android.support.v7.util.BatchingListUpdateCallback.dispatchLastEvent(BatchingListUpdateCallback.java:64)
        at android.support.v7.util.BatchingListUpdateCallback.onInserted(BatchingListUpdateCallback.java:82)
        at android.support.v7.util.DiffUtil$DiffResult.dispatchAdditions(DiffUtil.java:881)
        at android.support.v7.util.DiffUtil$DiffResult.dispatchUpdatesTo(DiffUtil.java:840)
        at android.support.v7.util.DiffUtil$DiffResult.dispatchUpdatesTo(DiffUtil.java:802)
        at org.wordpress.android.ui.stats.refresh.lists.sections.BlockListAdapter.update(BlockListAdapter.kt:89)
        at org.wordpress.android.ui.stats.refresh.StatsViewAllFragment.loadData(StatsViewAllFragment.kt:206)
        at org.wordpress.android.ui.stats.refresh.StatsViewAllFragment.access$loadData(StatsViewAllFragment.kt:44)
        at org.wordpress.android.ui.stats.refresh.StatsViewAllFragment$setupObservers$3.onChanged(StatsViewAllFragment.kt:179)
        at org.wordpress.android.ui.stats.refresh.StatsViewAllFragment$setupObservers$3.onChanged(StatsViewAllFragment.kt:44)
        at android.arch.lifecycle.LiveData.considerNotify(LiveData.java:109)
        at android.arch.lifecycle.LiveData.dispatchingValue(LiveData.java:126)
        at android.arch.lifecycle.LiveData.setValue(LiveData.java:282)
        at android.arch.lifecycle.MutableLiveData.setValue(MutableLiveData.java:33)
        at org.wordpress.android.util.LiveDataUtilsKt$distinct$1.onChanged(LiveDataUtils.kt:180)
        at android.arch.lifecycle.MediatorLiveData$Source.onChanged(MediatorLiveData.java:152)
        at android.arch.lifecycle.LiveData.considerNotify(LiveData.java:109)
        at android.arch.lifecycle.LiveData.dispatchingValue(LiveData.java:126)
        at android.arch.lifecycle.LiveData.setValue(LiveData.java:282)
        at android.arch.lifecycle.MutableLiveData.setValue(MutableLiveData.java:33)
        at org.wordpress.android.util.LiveDataUtilsKt$map$1.onChanged(LiveDataUtils.kt:162)
        at android.arch.lifecycle.MediatorLiveData$Source.onChanged(MediatorLiveData.java:152)
        at android.arch.lifecycle.LiveData.considerNotify(LiveData.java:109)
        at android.arch.lifecycle.LiveData.dispatchingValue(LiveData.java:126)
        at android.arch.lifecycle.LiveData.setValue(LiveData.java:282)
        at android.arch.lifecycle.MutableLiveData.setValue(MutableLiveData.java:33)
        at org.wordpress.android.util.LiveDataUtilsKt$distinct$1.onChanged(LiveDataUtils.kt:180)
        at android.arch.lifecycle.MediatorLiveData$Source.onChanged(MediatorLiveData.java:152)
        at android.arch.lifecycle.LiveData.considerNotify(LiveData.java:109)
        at android.arch.lifecycle.LiveData.dispatchingValue(LiveData.java:126)
        at android.arch.lifecycle.LiveData.setValue(LiveData.java:282)
        at android.arch.lifecycle.MutableLiveData.setValue(MutableLiveData.java:33)
        at org.wordpress.android.util.LiveDataUtilsKt$map$1.onChanged(LiveDataUtils.kt:162)
        at android.arch.lifecycle.MediatorLiveData$Source.onChanged(MediatorLiveData.java:152)
        at android.arch.lifecycle.LiveData.considerNotify(LiveData.java:109)
2019-03-04 20:49:06.992 18752-18752/org.wordpress.android.beta E/AndroidRuntime:     at android.arch.lifecycle.LiveData.dispatchingValue(LiveData.java:126)
        at android.arch.lifecycle.LiveData.setValue(LiveData.java:282)
        at android.arch.lifecycle.MutableLiveData.setValue(MutableLiveData.java:33)
        at org.wordpress.android.util.LiveDataUtilsKt$merge$6.onChanged(LiveDataUtils.kt:127)
        at android.arch.lifecycle.MediatorLiveData$Source.onChanged(MediatorLiveData.java:152)
        at android.arch.lifecycle.LiveData.considerNotify(LiveData.java:109)
        at android.arch.lifecycle.LiveData.dispatchingValue(LiveData.java:126)
        at android.arch.lifecycle.LiveData.setValue(LiveData.java:282)
        at android.arch.lifecycle.MutableLiveData.setValue(MutableLiveData.java:33)
        at org.wordpress.android.ui.stats.refresh.lists.sections.BaseStatsUseCase.updateUiState(BaseStatsUseCase.kt:125)
        at org.wordpress.android.ui.stats.refresh.lists.sections.insights.usecases.FollowersUseCase.loadMore(FollowersUseCase.kt:158)
        at org.wordpress.android.ui.stats.refresh.lists.sections.insights.usecases.FollowersUseCase.access$loadMore(FollowersUseCase.kt:46)
        at org.wordpress.android.ui.stats.refresh.lists.sections.insights.usecases.FollowersUseCase$buildStatefulUiModel$3.invoke(FollowersUseCase.kt:148)
        at org.wordpress.android.ui.stats.refresh.lists.sections.insights.usecases.FollowersUseCase$buildStatefulUiModel$3.invoke(FollowersUseCase.kt:46)
        at org.wordpress.android.ui.stats.refresh.lists.sections.viewholders.LoadingItemViewHolder.bind(LoadingItemViewHolder.kt:13)
        at org.wordpress.android.ui.stats.refresh.lists.sections.BlockListAdapter.onBindViewHolder(BlockListAdapter.kt:147)
        at org.wordpress.android.ui.stats.refresh.lists.sections.BlockListAdapter.onBindViewHolder(BlockListAdapter.kt:67)
        at android.support.v7.widget.RecyclerView$Adapter.bindViewHolder(RecyclerView.java:6823)
        at android.support.v7.widget.RecyclerView$Recycler.tryBindViewHolderByDeadline(RecyclerView.java:5752)
        at android.support.v7.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline(RecyclerView.java:6019)
        at android.support.v7.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:5858)
        at android.support.v7.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:5854)
        at android.support.v7.widget.LinearLayoutManager$LayoutState.next(LinearLayoutManager.java:2230)
        at android.support.v7.widget.LinearLayoutManager.layoutChunk(LinearLayoutManager.java:1557)
        at android.support.v7.widget.LinearLayoutManager.fill(LinearLayoutManager.java:1517)
        at android.support.v7.widget.LinearLayoutManager.onLayoutChildren(LinearLayoutManager.java:612)
        at android.support.v7.widget.RecyclerView.dispatchLayoutStep2(RecyclerView.java:3924)
        at android.support.v7.widget.RecyclerView.dispatchLayout(RecyclerView.java:3641)
        at android.support.v7.widget.RecyclerView.onLayout(RecyclerView.java:4194)
        at android.view.View.layout(View.java:20672)
        at android.view.ViewGroup.layout(ViewGroup.java:6194)
        at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
        at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
        at com.facebook.shimmer.ShimmerFrameLayout.onLayout(ShimmerFrameLayout.java:105)
        at android.view.View.layout(View.java:20672)
        at android.view.ViewGroup.layout(ViewGroup.java:6194)
        at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1812)
        at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1656)
        at android.widget.LinearLayout.onLayout(LinearLayout.java:1565)
        at android.view.View.layout(View.java:20672)
        at android.view.ViewGroup.layout(ViewGroup.java:6194)
        at android.support.v4.widget.SwipeRefreshLayout.onLayout(SwipeRefreshLayout.java:625)
        at android.view.View.layout(View.java:20672)
        at android.view.ViewGroup.layout(ViewGroup.java:6194)
        at android.support.design.widget.HeaderScrollingViewBehavior.layoutChild(HeaderScrollingViewBehavior.java:142)
        at android.support.design.widget.ViewOffsetBehavior.onLayoutChild(ViewOffsetBehavior.java:41)
2019-03-04 20:49:06.993 18752-18752/org.wordpress.android.beta E/AndroidRuntime:     at android.support.design.widget.AppBarLayout$ScrollingViewBehavior.onLayoutChild(AppBarLayout.java:1556)
        at android.support.design.widget.CoordinatorLayout.onLayout(CoordinatorLayout.java:888)
        at android.view.View.layout(View.java:20672)
        at android.view.ViewGroup.layout(ViewGroup.java:6194)
        at android.support.constraint.ConstraintLayout.onLayout(ConstraintLayout.java:1858)
        at android.view.View.layout(View.java:20672)
        at android.view.ViewGroup.layout(ViewGroup.java:6194)
        at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1812)
        at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1656)
        at android.widget.LinearLayout.onLayout(LinearLayout.java:1565)
        at android.view.View.layout(View.java:20672)
        at android.view.ViewGroup.layout(ViewGroup.java:6194)
        at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
        at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
        at android.view.View.layout(View.java:20672)
        at android.view.ViewGroup.layout(ViewGroup.java:6194)
        at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
        at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
        at android.view.View.layout(View.java:20672)
        at android.view.ViewGroup.layout(ViewGroup.java:6194)
        at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
        at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
        at android.view.View.layout(View.java:20672)
        at android.view.ViewGroup.layout(ViewGroup.java:6194)
        at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1812)
        at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1656)
        at android.widget.LinearLayout.onLayout(LinearLayout.java:1565)
        at android.view.View.layout(View.java:20672)
        at android.view.ViewGroup.layout(ViewGroup.java:6194)
        at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
        at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
        at com.android.internal.policy.DecorView.onLayout(DecorView.java:753)
        at android.view.View.layout(View.java:20672)
        at android.view.ViewGroup.layout(ViewGroup.java:6194)
        at android.view.ViewRootImpl.performLayout(ViewRootImpl.java:2792)
        at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2319)
        at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1460)
        at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:7183)
        at android.view.Choreographer$CallbackRecord.run(Choreographer.java:949)
        at android.view.Choreographer.doCallbacks(Choreographer.java:761)
        at android.view.Choreographer.doFrame(Choreographer.java:696)
        at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:935)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6669)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
```

**Comment by planarvoid** [comment]
@0nko that should be fixed now, thanks for noticing! there was one missing change

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/insights/usecases/FollowersUseCase.kt [hunk 1]
```diff
@@ -119,35 +128,49 @@ class FollowersUseCase
...
         } else {
             items.add(Empty())
         }
         return items
     }
 
+    private fun loadMore() {
+        GlobalScope.launch(bgDispatcher) {
+            withContext(mainDispatcher) {
+                updateUiState { it.copy(isLoading = true) }
```
Comment (0nko) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/insights/usecases/FollowersUseCase.kt]: Is this still necessary now that the UI state is updated inside `fetchData()`?

**Comment by designsimply** [comment]
I tested with Pixel 2, Android 9.0 (emulator) and infinite scroll worked well ([10s](https://cloudup.com/csRy0Z7pp00)). 👍

**Review by designsimply — COMMENTED** [review]
I tested with Pixel 2, Android 9.0 (emulator) and infinite scroll worked well ([10s](https://cloudup.com/csRy0Z7pp00)). 👍

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/insights/usecases/FollowersUseCase.kt (hunk 1 above)
Comment (planarvoid) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/insights/usecases/FollowersUseCase.kt]: ah you're right, I didn't realise `loadMore` calls `fetchData`, good call 👍

**Comment by 0nko** [comment]
Looks good, thanks. There are just some tests that need to be updated.

**Comment by 0nko** [comment]
Works great! Thanks for all the changes 🎖

**Pull request merged**